util/http.cc

```diff
@@ -285,12 +291,131 @@ int http::post(const std::string& url, const char* content_type,
                  compressed_payload.get(), compressed_size);
 }
 
-static constexpr const char* const json_type = "Content-Type: application/json";
 int http::post(const std::string& url,
                const rapidjson::Document& payload) const {
   rapidjson::StringBuffer buffer;
   auto c_str = JsonGetString(buffer, payload);
-  return post(url, json_type, c_str, strlen(c_str));
+  return post(url, kJsonType, c_str, std::strlen(c_str));
+}
+
+class CurlMultiHandle {
+ public:
+  CurlMultiHandle() : multi_handle_(curl_multi_init()) {}
+  CurlMultiHandle(const CurlMultiHandle&) = delete;
+  CurlMultiHandle(CurlMultiHandle&&) = delete;
+  CurlMultiHandle& operator=(const CurlMultiHandle&) = delete;
+  CurlMultiHandle& operator=(CurlMultiHandle&&) = delete;
+  ~CurlMultiHandle() { curl_multi_cleanup(multi_handle_); }
+
+  std::vector<int> perform_all(
+      const std::vector<std::unique_ptr<CurlHandle>>& easy_handles) {
+    for (const auto& h : easy_handles) {
+      curl_multi_add_handle(multi_handle_, h->handle());
+    }
+
+    auto still_running = 0;
+    auto repeats = 0;
+    do {
+      auto mc = curl_multi_perform(multi_handle_, &still_running);
+      auto numfds = 0;
+      if (mc == CURLM_OK) {
+        // wait for activity, timeout, or "nothing"
+        mc = curl_multi_wait(multi_handle_, nullptr, 0, 1000, &numfds);
+      }
+      if (mc != CURLM_OK) {
+        Logger()->warn("curl_multi failed: {}", curl_multi_strerror(mc));
+        break;
+      }
+
+      // 'numfds' being zero means either a timeout or no file descriptors to
+      // wait for. Try timeout on first occurrence, then assume no file
+      // descriptors and no file descriptors to wait for means wait for 100
+      // milliseconds.
+      if (numfds == 0) {
+        repeats++;  // count number of repeated zero numfds
+        if (repeats > 1) {
+          std::this_thread::sleep_for(std::chrono::milliseconds(100));
+        }
+      } else {
+        repeats = 0;
+      }
+    } while (still_running > 0);
+
+    CURLMsg* msg; /* for picking up messages with the transfer status */
+    int msgs_left;
+    std::vector<int> res(easy_handles.size());
+    std::fill(res.begin(), res.end(), 400);
+    while ((msg = curl_multi_info_read(multi_handle_, &msgs_left)) != nullptr) {
+      if (msg->msg == CURLMSG_DONE) {
+        // Find out which handle this message is about
+        for (auto i = 0u; i < easy_handles.size(); i++) {
+          auto easy_handle = easy_handles[i]->handle();
+          bool found = (msg->easy_handle == easy_handle);
+          if (found) {
+            auto easy_code = msg->data.result;
+            if (easy_code == CURLE_OK) {
+              res.at(i) = easy_handles[i]->status_code();
+            } else {
+              Logger()->warn("Failed batch: {}", curl_easy_strerror(easy_code));
+              res.at(i) = 400;
```

Comment:
Not sure what uses this status, but using the curl error code would give us the ability to diagnose the cause from:

https://curl.haxx.se/libcurl/c/libcurl-errors.html

May not matter much since the above warning log would probably be easier though.

Reply:
It seems we could use either a number < 100 to indicate a curl error code, or we could add a big constant to distinguish between curl error codes and http status codes.